### PR TITLE
Gh 18 redux

### PR DESCRIPTION
**Pull request by evdhoudt**
Closes  NCI-Agency/anet#18

### CODE REVIEW

**Comment by evdhoudt** [comment]
At the moment I've implemented a redux solution to remember pagination for my reports and organization. 

However, I was wondering if this should also be implemented for all search pagination field sets? 

Also, could you check if the behaviour is as expected. At the moment the pagination is remembered throughout the session. I was thinking that it might be useful that the redux store is cleared when a link in the sidebar menu or the main navigation is clicked. What do you think?

**Comment by evdhoudt** [comment]
The branch is ready for review :)

File: client/src/components/LinkTo.js [hunk 1]
```diff
@@ -16,6 +16,15 @@ const MODEL_NAMES = Object.keys(Models).map(key => {
...
 
+const modelPropTypes = MODEL_NAMES.reduce((map, name) => ({
+	...map,
+	[name]: PropTypes.oneOfType([
+		PropTypes.instanceOf(Models[name]),
+		PropTypes.object,
+		PropTypes.string,
+	])
+}), {})
+
```
Comment (gjvoosten) [inline client/src/components/LinkTo.js]: How did this change from `candidate-2.1` end up in this PR?

File: client/src/components/Page.js [hunk 2]
```diff
@@ -12,7 +12,16 @@ import utils from 'utils'
...
-import { setPageProps, setSearchProps, setSearchQuery, clearSearchQuery, DEFAULT_PAGE_PROPS, DEFAULT_SEARCH_PROPS} from 'actions'
+import {
+	setPageProps,
+	setSearchProps,
+	setSearchQuery,
+	clearSearchQuery,
+	DEFAULT_PAGE_PROPS,
+	DEFAULT_SEARCH_PROPS,
+	goToPaginated,
+	resetPages,
```
Comment (gjvoosten) [inline client/src/components/Page.js]: Move `goToPaginated` and `resetPages` to between `clearSearchQuery` and `DEFAULT_PAGE_PROPS`.

File: client/src/components/TopBar.js [hunk 3]
```diff
@@ -109,4 +125,9 @@ const TopBar = (props) => (
 	</AppContext.Consumer>
 )
 
-export default TopBar
+export const mapDispatchToProps = (dispatch) => ({
```
Comment (gjvoosten) [inline client/src/components/TopBar.js]: Removed unneeded `export`.

File: client/src/components/TopBar.js [hunk 4]
```diff
@@ -109,4 +125,9 @@ const TopBar = (props) => (
...
+	resetPages: () => dispatch(resetPages()),
+})
+
+
+export default  connect(null, mapDispatchToProps)(TopBar)
```
Comment (gjvoosten) [inline client/src/components/TopBar.js]: Remove extra space.

File: client/src/reducers/pagination.test.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
+import reducer from './pagination'
```
Comment (gjvoosten) [inline client/src/reducers/pagination.test.js]: Should this file be under `client/tests`?

File: client/src/store/configureStore.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import { createStore, compose } from 'redux'
+import { persistStore } from 'redux-persist'
+//import { reduxReactRouter, routerStateReducer, ReduxRouter } from 'redux-router'
+//import { createHashHistory as history } from 'history'
```
Comment (gjvoosten) [inline client/src/store/configureStore.js]: Remove commented out imports.

File: client/src/store/configureStore.js [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+import persistReducer from '../reducers/persistReducer'
+
+export default function configureStore(initialState) {
+  let finalCreateStore = compose(
+	//reduxReactRouter({ history }),
```
Comment (gjvoosten) [inline client/src/store/configureStore.js]: Remove commented out code.

File: client/src/store/configureStore.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import { createStore, compose } from 'redux'
+import { persistStore } from 'redux-persist'
+//import { reduxReactRouter, routerStateReducer, ReduxRouter } from 'redux-router'
+//import { createHashHistory as history } from 'history'
+import persistReducer from '../reducers/persistReducer'
+
+export default function configureStore(initialState) {
```
Comment (gjvoosten) [inline client/src/store/configureStore.js]: Document what this is supposed to do.

File: client/src/pages/Search.js [hunk 9]
```diff
@@ -519,12 +529,13 @@ class Search extends Page {
 
 	@autobind
 	exportSearchResults() {
-		this._dataFetcher(this.props, this._exportSearchResultsCallback, 0, 0)
+		this._dataFetcher(this.props, this._exportSearchResultsCallback, 0)
```
Comment (gjvoosten) [inline client/src/pages/Search.js]: Why was this change from `candidate-2.1` reverted?

File: client/src/pages/organizations/Show.js [hunk 10]
```diff
@@ -325,13 +331,20 @@ class BaseOrganizationShow extends Page {
 		)
 	}
 
+	orgLabel = (organization) => {
+		const shortName = organization.shortName
+		const label = shortName.replace(/\s/g, '')
+		return `reports_${label}`
+	}
```
Comment (maradragan) [inline client/src/pages/organizations/Show.js]: I think this makes the orgLabel not unique. What if you have 2 organizations with the same shortName and view them after eachother? I think the last pagination will be used when going back (but I didn't try it)?

File: client/src/components/LinkTo.js (hunk 1 above)
Comment (evdhoudt) [reply client/src/components/LinkTo.js]: Removed from PR

Comment (evdhoudt) [reply client/src/components/LinkTo.js]: Removed from PR

File: client/src/actions/index.js [hunk 11]
```diff
@@ -55,3 +57,19 @@ export const setSearchQuery = searchQuery => ({
...
+	return {
+		type: types.GOTO_PAGE,
+		payload: {
+			pageKey,
+			pageNum
+		}
+	}
+}
+
+export function resetPages() {
```
[the remark this answers is not in the document]
Comment (evdhoudt) [reply client/src/actions/index.js]: I did not use arrow functions on purpose.
I prefer vanilla JS functions to arrow functions when scope/context for this is not required. I'm not sure if we have a convention for this. But I agree, either way we should change all actions to be the same type.

File: client/src/reducers/pagination.test.js (hunk 5 above)
Comment (evdhoudt) [reply client/src/reducers/pagination.test.js]: For now I've added the according reducer and action tests in the client/tests folder.
However, when keeping the tests close to the file it is easier to quickly see which files do NOT have any tests yet.
Could be interesting to take this in to account in the future

File: client/src/pages/Search.js (hunk 9 above)
Comment (evdhoudt) [reply client/src/pages/Search.js]: I was too eager to introduce the redux paginated pageNum everywhere.
The export function obviously does not need paginated results.

**Comment by evdhoudt** [comment]
@gjvoosten @maradragan @VassilIordanov
Pushed new commits, please review :)

File: client/src/pages/Search.js [hunk 12]
```diff
@@ -121,16 +122,30 @@ class Search extends Page {
...
 
+	pageLabel = (type, prefix = this.componentPrefix) => {
+		return `${prefix}${type}`
+	}
+
+	getSearchPart(type, query, pageNum = 0, pageSize = 10) {
+		const typeLower = type.toLowerCase()
 		let subQuery = Object.assign({}, query)
-		subQuery.pageNum = (paginatedPart === undefined) ? pageNum : paginatedPart.pageNum
+		subQuery.pageNum = pageNum //(paginatedPart === undefined) ? pageNum : paginatedPart.pageNum
```
Comment (maradragan) [inline client/src/pages/Search.js]: Please remove the commented code at the end of the line.

File: client/src/pages/Search.js [hunk 13]
```diff
@@ -121,16 +122,30 @@ class Search extends Page {
...
 		const typeLower = type.toLowerCase()
 		const pageLabel = this.pageLabel(typeLower)
-		const paginatedPart = pagination[pageLabel]
+		return  pagination[pageLabel]
+	}
 
+	pageLabel = (type, prefix = this.componentPrefix) => {
+		return `${prefix}${type}`
+	}
+
```
Comment (maradragan) [inline client/src/pages/Search.js]: The commit mentions that it improves the paginated labels, but I don't see where that is done.
Also, it mentions it fixes the paginated reports, but what was the problem?

File: client/src/components/SearchBar.js [hunk 14]
```diff
@@ -92,6 +93,7 @@ class SearchBar extends Component {
 	onSubmit(event) {
 		if (!this.state.showAdvancedSearch) {
 			// We only update the Redux state on submit
+			this.props.resetPages()
```
Comment (maradragan) [inline client/src/components/SearchBar.js]: I think this is why now we have the following problem:
- I search on positions (also add some filters on positions)
- I go to the search bar, change the search term and click search
- the search filter is gone now

Comment (evdhoudt) [reply client/src/components/SearchBar.js]: Should be fixed

File: client/src/pages/Search.js (hunk 12 above)
Comment (evdhoudt) [reply client/src/pages/Search.js]: Done

File: client/src/pages/Search.js (hunk 13 above)
Comment (evdhoudt) [reply client/src/pages/Search.js]: This commit should be:

GH-18: Fix pagination in Search and MyReports

Search should use different pagination variables with regard to reports.
MyReports pagination was broken and is fixed by providing the get part with the correct page number.

@gjvoosten @maradragan I'm not sure how to fix the commit message. Should I rebase the commit message and force push?

Comment (maradragan) [reply client/src/pages/Search.js]: @evdhoudt yes

Comment (evdhoudt) [reply client/src/pages/Search.js]: @maradragan I discussed this with @gjvoosten last week and we agreed that rebasing should not be done on this branch anymore. So... the only way to fix the message is to cherry-pick all commits to a new branch and rebase it there... This is a lot of work and I'm not sure if it is worth the effort.

File: client/package.json [hunk 15]
```diff
@@ -44,7 +44,7 @@
     "html-loader": "^0.5.5",
     "html-webpack-plugin": "^3.0.7",
     "ignore-loader": "^0.1.2",
-    "jest": "^24.1.0",
+    "jest": "^23.6.0",
```
Comment (gjvoosten) [inline client/package.json]: Should this change be here?

File: client/yarn.lock [hunk 16]
```diff
@@ -42,7 +42,7 @@
     imurmurhash "^0.1.4"
     slide "^1.1.5"
 
-"@babel/code-frame@7.0.0", "@babel/code-frame@^7.0.0":
+"@babel/code-frame@7.0.0", "@babel/code-frame@^7.0.0", "@babel/code-frame@^7.0.0-beta.35":
```
Comment (gjvoosten) [inline client/yarn.lock]: If there are no changes to `package.json`, there should be no changes to `yarn.lock`.

**Pull request merged**